Rework asynchronous fetches from SHAMap::getMissingNodes to Database::threadEntry

JoelKatz opened the pull request:
## High Level Overview of Change

This change reworks the way the node database handles asynchronous fetches, eliminating the positive and negative caches and allowing the reads to be processed as they complete rather than waiting for the last one to finish before processing the first one.

### Context of Change

The read generation logic was implemented some time ago and has always been a bit kludgy. It had important edge cases it mishandled, such as the case where an async read got pushed out of memory before it was retrieved. This removes that logic and replaces it with a callback scheme.

## Before / After

Prior to this change, asynchronous fetches were batched into generations with the results being placed into the database layer's positive and negative caches. The get missing nodes logic then retrieved the results from the caches. After this change, asynchronous fetches include a callback function that is called as soon as the operation completes. There is no need for the caches. The get missing nodes logic processes the reads as they complete.

## Future Tasks

Pushing the callback logic all the way into the NuDB layer and eliminating the Database layer's read threads would be a huge win. Every operation system we support has some form of async I/O and this would allow a larger I/O queue depth, improving performance on modern SSDs and also reducing context switches.

## Review notes

Some of the types are annoyingly verbose. It may be considered an improvement to replace them with `using` declarations or other trickery (like `::mapped_type` or something like that). Improvements welcomed.

JoelKatz committed:
[FOLD] Remove dead code

This commit is separate for easier review. It should be folded
into the redo_async commit.

This removes the batch fetch logic in the nodestore's Backend
class. It was never supported or used.

JoelKatz committed:
Rework deferred node logic and async fetch behavior

This comment explains this patch and the associated patches
that should be folded into it. This paragraph should be removed
when the patches are folded after review.

This change significantly improves ledger sync and fetch
times while reducing memory consumption. The change affects
the code from that begins with SHAMap::getMissingNodes and runs
through to Database::threadEntry.

The existing code issues a number of async fetches which are then
handed off to the Database's pool of read threads to execute.
The results of each read are placed in the Database's positive
and negative caches. The caller waits for all reads to complete
and then retrieves the results out of these caches.

Among other issues, this means that the results of the first read
cannot be processed until the last read completes. Additionally,
all the results must sit in memory.

This patch changes the behavior so that each read operation has a
completion handler associated with it. The completion of the read
calls the handler, allowing the results of each read to be
processed as it completes. As this was the only reason the
negative and positive caches were needed, they can now be removed.

The read generation code is also no longer needed and is removed.
The batch fetch logic was never implemented or supported and is
removed.

JoelKatz committed:
[FOLD] Remove read generation

This commit is separate for easier review. It should be folded
into the redo_async commit.

Remove the read generation logic from the Database class. It
is made obsolete by the new async logic.

JoelKatz committed:
[FOLD] Remove caches

This commit is separate for easier review. It should be folded
into the redo_async commit.

Remove the NodeStore's positive and negative caches. They are
no longer needed.

seelabs commented:
We started formatting the code with clang-format. Here's a patch to cherry pick that's be formatted: https://github.com/seelabs/rippled/commit/796a45f16caf6dbbd44fed62affac091bad06dc7

nbougalis commented on src/ripple/shamap/impl/SHAMap.cpp:
```diff
@@ -147,33 +147,40 @@ SHAMap::findKey(uint256 const& id) const
+        return {};
     }
 
+    std::shared_ptr<SHAMapAbstractNode> node;
+    try
+    {
+        node = SHAMapAbstractNode::makeFromPrefix(
+            makeSlice(object->getData()), hash);
+        if (node)
+            canonicalize(hash, node);
+    }
+    catch (std::exception const&)
+    {
+        JLOG(journal_.warn()) << "Invalid DB node " << hash;
+        return std::shared_ptr<SHAMapTreeNode>();
+    }
     return node;
```
Micro-nit. Recommend:

```
    try
    {
        auto node = SHAMapAbstractNode::makeFromPrefix(
            makeSlice(object->getData()), hash);
        if (node)
            canonicalize(hash, node);
        return node;
    }
    catch (std::exception const&)
    {
        JLOG(journal_.warn()) << "Invalid DB node " << hash;
        return {};
    }
```

nbougalis commented on src/ripple/shamap/impl/SHAMapSync.cpp:
```diff
@@ -327,7 +323,7 @@ SHAMap::getMissingNodes(int max, SHAMapSyncFilter* filter)
         max,
         filter,
-        f_.db().getDesiredAsyncReadCount(ledgerSeq_),
+        4096,   // number of async reads per pass
```
Should this number be tunable? Is it possible that 4096 will be too much (or too little)?

JoelKatz replied:
There's really no basis on which to tune it. It controls how quickly getMissingNodes switches between looking for more async reads to do and waiting for async reads to complete. There's no reason to think more is better, you just need a high enough number that the cost of waiting for reads to complete is lost in the noise. You just want to stay in the "discovering reads we need to do" phase long enough that a couple of the reads you've pended have completed to minimize the time you're doing nothing.

nbougalis replied:
:+1:

nbougalis commented on src/ripple/nodestore/impl/DatabaseNodeImp.cpp:
```diff
@@ -77,8 +61,8 @@ DatabaseNodeImp::fetchNodeObject(
     FetchReport& fetchReport)
 {
     // See if the node object exists in the cache
```
This comment is no longer applicable and should be removed.

nbougalis reviewed (commented):
Looks good. Left a few comments, mostly stylistic. Happy to do do those and squash this down into one commit for you.

JoelKatz commented on src/ripple/nodestore/impl/Database.cpp:
```diff
@@ -312,19 +290,20 @@ Database::threadEntry()
 
-        // Perform the read
-        fetchNodeObject(lastHash, lastSeq, FetchType::async);
+        for (auto const& req : entry)
+        {
+            auto obj = fetchNodeObject(lastHash, req.first, FetchType::async);
+            req.second (obj);
+        }
```
I think that won't work since you can't bind a non-const reference to a temporary.

nbougalis replied:
Yes, but I think that the callback can (and, given how it's used _should_) be changed to take `std::shared_ptr<NodeObject> const&` or, perhaps better still, simply `NodeObject*`. Sorry I wasn't more clear.

JoelKatz replied:
I made this change, but it forces an extra copy. You can move both to and from a parameter passed by value. When you pass a parameter by reference, the recipient cannot move from it and must make a deep copy.

seelabs committed:
[fold] Reformat with clang-format

seelabs committed:
[fold] refactor asyncFetch interface

seelabs commented on src/ripple/nodestore/Database.h:
```diff
@@ -135,20 +135,23 @@ class Database : public Stoppable
-        std::shared_ptr<NodeObject>& nodeObject) = 0;
+        std::shared_ptr<NodeObject>& nodeObject,
+        std::function<void(
+            std::shared_ptr<NodeObject>&)>&& callback) = 0;
```
I'd like to refactor this interface. `nodeObject` is never used, we always get the `nodeObject` through the callback. All implementations always post a read, _except_ `shards`, which first checks that the ledger is part  of its collection. If it's not, it doesn't post a read, and will not retrieve the object. Given that, I'd like to get rid of the `nodeObject` parameter, and make the return value `true` if pending (i.e. "success") and `false` if the object will never be found (failure).

Here's a patch for that refactor: https://github.com/seelabs/rippled/commit/f272b74741a8bc9d0e51f279fa978f9ed4a8ecf9 

Two notes: 
* this patch is based on top of a clang-reformat patch. You need the reformat-patch before applying this patch).
* this patch contains a fix for `DatabaseShardImp::asyncFetch` that I describe in another comment

seelabs replied:
It looks like `DatabaseShardImp::asyncFetch` can be reomved completely. In that case, this function can now return `void`. Here's a second patch that builds on the first: https://github.com/ripple/rippled/commit/6147739d1a71b522597b632988fb055170c707dc

seelabs commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -1027,16 +1028,13 @@ DatabaseShardImp::asyncFetch(
-            return false;
+            return true;
         shard = it->second;
     }
```
This function doesn't look correct. We check if the `acquireIndex_` is in shards and we don't fetch if it's not? That doesn't make sense to me. I spoke with @miguelportilla and he agrees that this looks buggy. He suggested the following fix:
```
bool
DatabaseShardImp::asyncFetch(
    uint256 const& hash,
    std::uint32_t ledgerSeq,
    std::function<void(std::shared_ptr<NodeObject>&)>&& callback)
{
    auto const shardIndex{seqToShardIndex(ledgerSeq)};
    {
        std::lock_guard lock(mutex_);
        assert(init_);
        if (shards_.find(shardIndex) == shards_.end())
            return false;
    }

    // Otherwise post a read
    return Database::asyncFetch(hash, ledgerSeq, std::move(callback));
}
```
Notes:
* this is part of the patch here: https://github.com/seelabs/rippled/commit/f272b74741a8bc9d0e51f279fa978f9ed4a8ecf9
* this uses the new return value semantics (true means pending, false mean it's never coming).

seelabs replied:
Actually, looking closer, `DatabaseShardImp::fetchNodeObject(` already does this check. This means we can get rid of all the overides. I'll make a patch to do so.

seelabs replied:
Here's a second patch that builds on the first: https://github.com/ripple/rippled/commit/6147739d1a71b522597b632988fb055170c707dc @miguelportilla Please dummy check that it is OK to remove `DatabaseShardImp::asyncFetch` and use `fetchNodeObject` to do this check for us.

miguelportilla replied:
The patches look good to me. The drawback is that we can now post a read knowing that the ledger sequence does not correspond to any shard. But it won't make it to IO as `DatabaseShardImp::fetchNodeObject` will return early (as you mentioned). This cleanup's benefit far outweighs that.

seelabs reviewed (commented):
Leaving a partial review for refactoring the `asyncFetch` interface. This patch has the suggestions: https://github.com/seelabs/rippled/commit/f272b74741a8bc9d0e51f279fa978f9ed4a8ecf9

Edit: Here's a second patch that builds on the first: https://github.com/ripple/rippled/commit/6147739d1a71b522597b632988fb055170c707dc

seelabs committed:
[fold] remove DatabaseShardImp::asyncFetch and always use default impl

seelabs commented on src/ripple/nodestore/impl/Database.cpp:
```diff
@@ -312,19 +290,20 @@ Database::threadEntry()
-        fetchNodeObject(lastHash, lastSeq, FetchType::async);
+        for (auto const& req : entry)
+        {
+            auto obj = fetchNodeObject(lastHash, req.first, FetchType::async);
```
How often do we expect to have multiple callbacks in the `entry` collection with the same ledgerSeq? When that happens we make redundant calls to `fetchNodeObject` that we don't need.

Two suggestions to address this:
* Sort the `entry` vector by the first element in the pair, and keep track of the last ledgerIndex used in `fetchNodeObject` and only call that function when `req.first` is different from the last call
* Instead of entry being a vector of pairs, it could be a `unordered_map<std::uint32_t,std::vector<std::function<std::shared_ptr<NodeObject>&)>>>>`

Of course, if we think this is a rare case it might not be worth handling it.

miguelportilla replied:
@seelabs It is redundant only when the ledger sequence and the node object hash are the same. My guess is that's unlikely but it should be examined closer.

@JoelKatz You are using the same node object hash for every fetch. You may need to make `entry` the same type as `read_` to retain the hash.

JoelKatz replied:
I think this is sufficiently unlikely to happen and the consequences of it happening (an extra read) so mild that I left the code as simple as possible.

JoelKatz replied:
Suppose the same node is in two different ledgers, each in a different shard. We may be searching for both nodes simultaneously. It is critical that we neither tell one query that the node is not in the shard if it is and also critical that we not tell one query that the node is in the shard if it isn't. Otherwise, we may build incomplete shards and fail to retrieve from complete ones!

seelabs commented on src/ripple/shamap/SHAMap.h:
```diff
@@ -478,6 +491,12 @@ class SHAMap
+    std::shared_ptr<SHAMapAbstractNode>
+    finishFetch(
+        SHAMapHash const& hash,
+        std::shared_ptr<NodeObject>& object) const;
```
`object` could be `const&`

seelabs commented on src/ripple/shamap/impl/SHAMap.cpp:
```diff
@@ -377,19 +385,29 @@ SHAMap::descendAsync(
-            std::shared_ptr<NodeObject> obj;
-            if (!f_.db().asyncFetch(hash.as_uint256(), ledgerSeq_, obj))
+            std::shared_ptr<NodeObject> object;
+            if (f_.db().asyncFetch(hash.as_uint256(), ledgerSeq_, object,
```
I'll note this is the only call to `asyncFetch`, which means all the calls use the same callback. We could consider removing the callback parameters and hard coding the callback - this makes the code easier to understand.

On the other hand, keeping the callbacks makes for a better interface to `Database` and keeps the components separate.

I'm in favour of keeping things as they are, but wanted to point this out in case other people had strong feelings of wanting to change this.

seelabs commented on src/ripple/shamap/impl/SHAMapSync.cpp:
```diff
@@ -202,25 +202,35 @@ SHAMap::gmn_ProcessNodes(MissingNodes& mn, MissingNodes::StackEntry& se)
             {
-                fullBelow = false;  // for now, not known full below
+                // a read completed asynchronously
+                std::unique_lock<std::mutex> lock{mn.deferLock_};
```
C++-17 can infer the template parameter. So this can be:
```
std::unique_lock lock{mn.deferLock_};
```

seelabs commented on src/ripple/shamap/impl/SHAMapSync.cpp:
```diff
@@ -259,30 +269,28 @@ SHAMap::gmn_ProcessNodes(MissingNodes& mn, MissingNodes::StackEntry& se)
+        std::tuple<SHAMapInnerNode*, SHAMapNodeID, int,
+            std::shared_ptr<SHAMapAbstractNode>> deferredNode;
+        {
+            std::unique_lock<std::mutex> lock{mn.deferLock_};
```
C++-17 can infer the template parameter. This can be just:
```
std::unique_lock lock{mn.deferLock_};
```

seelabs reviewed (commented):
*love* these changes. It simplifies existing code, uses less memory, and gives a performance boost. What's not to love?

Left a few more comments for you to consider.

miguelportilla commented on src/ripple/nodestore/impl/Database.cpp:
```diff
@@ -296,7 +301,9 @@ Database::threadEntry()
-        std::uint32_t lastSeq;
+        std::vector< std::pair <std::uint32_t,
+            std::function<void(std::shared_ptr<NodeObject>&)>>> entry;
+
```
I believe we can eliminate `lastHash` and move the `entry` declaration outside the loop.
eg.
```
    std::vector<std::pair<
        std::uint32_t,
        std::function<void(std::shared_ptr<NodeObject>&)>>>
        entry;
    while (true)
    {
        {
            std::unique_lock<std::mutex> lock(readLock_);
            while (!readShut_ && read_.empty())
            {
                // All work is done
                readCondVar_.wait(lock);
            }
            if (readShut_)
                break;

            // Read in key order to make the back end more efficient
            auto it = read_.lower_bound(readLastHash_);
            if (it == read_.end())
            {
                // start over from the beginning
                it = read_.begin();
            }
            readLastHash_ = it->first;
            entry = std::move(it->second);
            read_.erase(it);
        }

        for (auto const& req : entry)
        {
            auto obj =
                fetchNodeObject(readLastHash_, req.first, FetchType::async);
```

JoelKatz committed:
[FOLD] suggestions from code review

miguelportilla commented:
`src/ripple/nodestore/impl/Tuning.h` can be removed from the project.

JoelKatz committed:
[FOLD] remove nodestore/impl/Tuning.h

nbougalis commented:
I squashed and rebased this set of commits (02a3269..35b86f0) against the latest beta, which made changes to the `SHAMap` code. You can find the result here: [a66e39be134788dee9d216b09f12475c3e9f789f](https://github.com/nbougalis/rippled/commit/a66e39be134788dee9d216b09f12475c3e9f789f) if you want to use it.

JoelKatz committed:
[FOLD] Compress queries to same underlying database

When fetching historical data, we frequently fetch from
two sequential ledgers at the same time. We don't want
to do two reads for those.
